Add support for icon-offset property functions

jellevoost opened the pull request:
In this PR I've added support for property functions on "icon-offset".

Note that is applied to the rotation of a symbol if this is set, as was the case when normally setting this for a whole layer.

It was not very clear if mapbox maintains the style-spec so I have not updated it, if you want me to I will do so.

benchmark | master 5db58df | master 258da4f
--- | --- | ---
**map-load** | 102 ms  | 120 ms 
**style-load** | 93 ms  | 91 ms 
**buffer** | 1,018 ms  | 1,047 ms 
**fps** | 58 fps  | 60 fps 
**frame-duration** | 3.4 ms, 0% > 16ms  | 3.7 ms, 0% > 16ms 
**query-point** | 0.90 ms  | 0.99 ms 
**query-box** | 58.19 ms  | 58.83 ms 
**geojson-setdata-small** | 7 ms  | 4 ms 
**geojson-setdata-large** | 267 ms  | 275 ms

lucaswoj commented on js/symbol/shaping.js:
```diff
@@ -186,11 +186,13 @@ function align(positionedGlyphs, justify, horizontalAlign, verticalAlign, maxLin
 }
 
-function shapeIcon(image, layout) {
+function shapeIcon(image, layout, layer, featureProperties) {
```
As of this diff, the `layout` argument is entirely unused by `shapeIcon`. Could you please remove it?

It may be cleaner to pass `iconOffset` to this function directly instead of `layer` and `featureProperties`. (["stamp coupling"](https://en.wikipedia.org/wiki/Coupling_(computer_programming)))

lucaswoj commented:
Thank you very much for this PR! We're excited to work with you to add data-driven styling support for the `icon-offset` property.

jellevoost commented:
I've added a test and refactored the code a little bit. Had some trouble to get headless-gl working on my machine, so I'm hoping the expected png is generated correctly (it seemed correct by eye).

Thanks for the pointers on how the PR process works!

lucaswoj commented:
Thanks for creating `mapbox-gl-style-spec` and `mapbox-gl-test-suite` PRs @jellevoost! 😄 

The next step is to link this branch with the branches in those PRs. 

You can see [how that was done in `package.json` in #3672](https://github.com/mapbox/mapbox-gl-js/pull/3672/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2L25). Note that you will need to change the SHA and the repository in `package.json`.

lucaswoj commented:
ref https://github.com/mapbox/mapbox-gl-test-suite/pull/191
ref https://github.com/mapbox/mapbox-gl-style-spec/pull/615

lucaswoj commented:
Rebased and wired up to branches within Mapbox at https://github.com/mapbox/mapbox-gl-js/pull/3729
